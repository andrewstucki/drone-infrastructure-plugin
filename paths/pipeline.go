package paths

type pipeline struct {
	Name    string                 `yaml:"name"`
	Kind    string                 `yaml:"kind,omitempty"`
	Steps   []*step                `yaml:"steps,omitempty"`
	Trigger conditions             `yaml:"trigger,omitempty"`
	Attrs   map[string]interface{} `yaml:",inline"`
}

func (p *pipeline) matchOrExclude(changedFiles []string) bool {
	if p.Trigger.Paths.HasIncludes() || p.Trigger.Paths.HasExcludes() {
		for _, f := range changedFiles {
			if p.Trigger.Paths.match(f) {
				// we want to run the pipeline
				return false
			}
		}
	}
	// if only Trigger.Paths is set, Trigger.Attrs will be unset, so it must be initialized
	if p.Trigger.Attrs == nil {
		p.Trigger.Attrs = make(map[string]interface{})
	}
	p.Trigger.Attrs["event"] = map[string][]string{"exclude": {"*"}}
	return true
}

func (p *pipeline) update(changedFiles []string) bool {
	if p.Kind != "pipeline" {
		return false
	}

	updated := false
	if p.matchOrExclude(changedFiles) {
		updated = true
	}
	for _, s := range p.Steps {
		if s.matchOrExclude(changedFiles) {
			updated = true
		}
	}
	return updated
}
