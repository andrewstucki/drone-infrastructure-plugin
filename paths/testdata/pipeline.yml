---
kind: pipeline
name: include

trigger:
  paths:
    include:
      - README.md

---
kind: pipeline
name: exclude

trigger:
  paths:
    exclude:
      - README.md
  event:
    - push
    - tag

---
kind: pipeline
name: exclude-include

trigger:
  paths:
    include:
      - README.md
    exclude:
      - README.md
  event:
    - push
    - tag

---
kind: pipeline
name: exclude-no-match

trigger:
  paths:
    exclude:
      - no-match.md
---
kind: pipeline
name: include-no-match

trigger:
  paths:
    include:
      - no-match.md
---
kind: secret
type: external
data:
  password: 'drone/docker#password'
  username: 'drone/docker#username'
