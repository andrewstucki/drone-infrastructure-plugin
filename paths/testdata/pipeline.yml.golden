name: include
kind: pipeline
trigger:
    paths:
        include:
          - README.md
---
name: exclude
kind: pipeline
trigger:
    paths:
        exclude:
          - README.md
    event:
        exclude:
          - '*'
---
name: exclude-include
kind: pipeline
trigger:
    paths:
        exclude:
          - README.md
        include:
          - README.md
    event:
        exclude:
          - '*'
---
name: exclude-no-match
kind: pipeline
trigger:
    paths:
        exclude:
          - no-match.md
---
name: include-no-match
kind: pipeline
trigger:
    paths:
        include:
          - no-match.md
    event:
        exclude:
          - '*'
---
name: ""
kind: secret
data:
    password: drone/docker#password
    username: drone/docker#username
type: external
